src/block_leaky_relu.rs

```diff
@@ -0,0 +1,180 @@
+        num_inputs: num_inputs,
+        alpha: 0.3, // TODO consider how to extract this and make configurable
+    });
+    let mut block_outputs = bg.add_node(block, vec![input])?;
+    assert_eq!(block_outputs.len(), 1);
+    Ok(block_outputs.pop().unwrap())
+}
+
+impl BlockTrait for BlockLeakyRELU {
+    fn as_any(&mut self) -> &mut dyn Any {
+        self
+    }
+
+    #[inline(always)]
+    fn forward_backward(
+        &mut self,
+        further_blocks: &mut [Box<dyn BlockTrait>],
+        fb: &feature_buffer::FeatureBuffer,
+        pb: &mut port_buffer::PortBuffer,
+        update: bool,
+    ) {
+        debug_assert!(self.output_offset != usize::MAX);
+        debug_assert!(self.input_offset != usize::MAX);
+        debug_assert!(self.num_inputs > 0);
+
+        unsafe {
+            for i in 0..self.num_inputs as usize {
+                let x = *pb.tape.get_unchecked_mut(self.input_offset + i);
+                if x < 0.0 {
+                    *pb.tape.get_unchecked_mut(self.output_offset + i) = self.alpha * x;
+                } else {
+                    *pb.tape.get_unchecked_mut(self.output_offset + i) = x;
+                }
+
+                if x <= 0.0 {
+                    *pb.tape.get_unchecked_mut(self.input_offset + i) = self.alpha;
+                } else {
+                    *pb.tape.get_unchecked_mut(self.input_offset + i) = 1.0;
+                }
+            }
+
+            block_helpers::forward_backward(further_blocks, fb, pb, update);
+
+            if update {
+                for i in 0..self.num_inputs as usize {
+                    let gradient = *pb.tape.get_unchecked(self.output_offset + i);
+                    *pb.tape.get_unchecked_mut(self.input_offset + i) *= gradient;
+                }
+            }
+        } // unsafe end
+    }
+
+    fn forward(
+        &self,
+        further_blocks: &[Box<dyn BlockTrait>],
+        fb: &feature_buffer::FeatureBuffer,
+        pb: &mut port_buffer::PortBuffer,
+    ) {
+        debug_assert!(self.output_offset != usize::MAX);
+        debug_assert!(self.input_offset != usize::MAX);
+        debug_assert!(self.num_inputs > 0);
+
+        unsafe {
+            for i in 0..self.num_inputs as usize {
+                let x = *pb.tape.get_unchecked_mut(self.input_offset + i);
+                if x < 0.0 {
+                    *pb.tape.get_unchecked_mut(self.output_offset + i) = self.alpha * x;
+                } else {
+                    *pb.tape.get_unchecked_mut(self.output_offset + i) = x;
+                }
+            }
+            block_helpers::forward(further_blocks, fb, pb);
+        } // unsafe end
+    }
+
+    fn allocate_and_init_weights(&mut self, _mi: &model_instance::ModelInstance) {}
+
+    fn get_num_output_values(&self, output: graph::OutputSlot) -> usize {
+        assert!(output.get_output_index() == 0);
+        return self.num_inputs;
+    }
+
+    fn get_num_output_slots(&self) -> usize {
+        1
```

Comment:
This method's name does not reflect the contents -> why is this not a constant of the object?

Reply:
this method is an implementation of the abstract method from the BlockTrait trait